brave/src/main/java/brave/secondary_sampling/SecondarySamplingExtractor.java

```diff
@@ -16,77 +16,59 @@
 import brave.propagation.Propagation.Getter;
 import brave.propagation.TraceContext.Extractor;
 import brave.propagation.TraceContextOrSamplingFlags;
-import brave.secondary_sampling.SecondarySamplingPolicy.Trigger;
-import java.util.LinkedHashMap;
-import java.util.Map;
 
-import static brave.secondary_sampling.SecondarySampling.parseParameters;
+import static brave.secondary_sampling.SecondarySampling.EXTRA_FACTORY;
 
 /**
- * This extracts the {@link SecondarySampling#fieldName sampling header}, and performs any TTL or
- * triggering logic based on the policy configured for this service.
+ * This extracts the {@link SecondarySampling#fieldName sampling header}, and parses it a list of
+ * {@link SecondarySamplingState}. For each extracted sampling key, TTL and sampling takes place if
+ * configured.
  */
 final class SecondarySamplingExtractor<C, K> implements Extractor<C> {
   final Extractor<C> delegate;
   final Getter<C, K> getter;
-  final SecondarySamplingPolicy policy;
+  final SecondarySampler sampler;
   final K samplingKey;
 
   SecondarySamplingExtractor(SecondarySampling.Propagation<K> propagation, Getter<C, K> getter) {
     this.delegate = propagation.delegate.extractor(getter);
     this.getter = getter;
-    this.policy = propagation.secondarySampling.policy;
+    this.sampler = propagation.secondarySampling.sampler;
     this.samplingKey = propagation.samplingKey;
   }
 
   @Override public TraceContextOrSamplingFlags extract(C carrier) {
     TraceContextOrSamplingFlags result = delegate.extract(carrier);
-
     String maybeValue = getter.get(carrier, samplingKey);
-    SecondarySampling.Extra extra = new SecondarySampling.Extra();
-    TraceContextOrSamplingFlags.Builder builder = result.toBuilder().addExtra(extra);
-    if (maybeValue == null) return builder.build();
+    if (maybeValue == null) return result;
 
+    SecondarySampling.Extra extra = EXTRA_FACTORY.create();
+    boolean sampledLocal = false;
     for (String entry : maybeValue.split(",", 100)) {
-      String[] nameParameters = entry.split(";", 100);
-      String samplingKey = nameParameters[0];
-
-      Map<String, String> parameters = nameParameters.length > 1
-        ? parseParameters(nameParameters)
-        : new LinkedHashMap<>();
-      if (updateParametersAndSample(samplingKey, parameters)) {
-        extra.sampledKeys.add(samplingKey);
-        builder.sampledLocal(); // this means data will be recorded
-      }
-      extra.samplingKeyToParameters.put(samplingKey, parameters);
+      SecondarySamplingState.Builder builder = SecondarySamplingState.parseBuilder(entry);
```

Comment:
Following up my previous [comment](https://github.com/openzipkin-contrib/zipkin-secondary-sampling/pull/7/files#r328250029).

Here is where the hook will be needed to let the sampler make a decision to start a trace by examining the request and set the `sampling_key`.

Reply:
please start using unit tests because you have been misunderstanding things when reading code.